### PR TITLE
Java: Extend String dataflow models

### PR DESCRIPTION
**Pull request by Marcono1234**
Extends the dataflow String models to cover more methods.

I am not that familiar with the CSV model yet, and also don't know how to use the tools to automatically generate test cases.
Additionally there are a few open questions (see comments).

Any feedback is appreciated, but also feel free to just pick this changes up and complete them, if that would be easiest for you.

### CODE REVIEW

File: java/ql/lib/semmle/code/java/frameworks/Strings.qll [hunk 1]
```diff
@@ -11,19 +11,28 @@ private class StringSummaryCsv extends SummaryModelCsv {
         "java.lang;String;false;concat;(String);;Argument[0];ReturnValue;taint",
         "java.lang;String;false;concat;(String);;Argument[-1];ReturnValue;taint",
         "java.lang;String;false;copyValueOf;;;Argument[0];ReturnValue;taint",
-        "java.lang;String;false;endsWith;;;Argument[-1];ReturnValue;taint",
```
Comment (Marcono1234) [inline java/ql/lib/semmle/code/java/frameworks/Strings.qll:14]: Have removed this because no other `boolean` returning methods are modelled.

File: java/ql/lib/semmle/code/java/frameworks/Strings.qll [hunk 2]
```diff
@@ -11,19 +11,28 @@ private class StringSummaryCsv extends SummaryModelCsv {
...
         "java.lang;String;false;formatted;(Object[]);;Argument[-1];ReturnValue;taint",
         "java.lang;String;false;formatted;(Object[]);;ArrayElement of Argument[0];ReturnValue;taint",
-        "java.lang;String;false;getChars;;;Argument[-1];Argument[2];taint",
         "java.lang;String;false;getBytes;;;Argument[-1];ReturnValue;taint",
+        "java.lang;String;false;getBytes;(int,int,byte[],int);;Argument[-1];Argument[2];taint",
+        "java.lang;String;false;getChars;;;Argument[-1];Argument[2];taint",
         "java.lang;String;false;indent;;;Argument[-1];ReturnValue;taint",
         "java.lang;String;false;intern;;;Argument[-1];ReturnValue;taint",
-        "java.lang;String;false;join;;;Argument[0..1];ReturnValue;taint",
+        "java.lang;String;false;join;;;Argument[0..1];ReturnValue;taint", // TODO: ArrayElement of Argument?
```
Comment (Marcono1234) [inline java/ql/lib/semmle/code/java/frameworks/Strings.qll:26]: Should this here and maybe also in other cases use `ArrayElement of Argument` instead? What is the difference between tracking taint form an array (or varargs?) compared to tracking taint from the elements?

File: java/ql/lib/semmle/code/java/frameworks/Strings.qll [hunk 3]
```diff
@@ -35,26 +44,43 @@ private class StringSummaryCsv extends SummaryModelCsv {
...
         "java.lang;String;false;toUpperCase;;;Argument[-1];ReturnValue;taint",
+        "java.lang;String;false;transform;;;Argument[-1];Parameter[0] of Argument[0];value",
+        "java.lang;String;false;transform;;;ReturnValue of Argument[0];ReturnValue;value",
         "java.lang;String;false;translateEscapes;;;Argument[-1];ReturnValue;taint",
         "java.lang;String;false;trim;;;Argument[-1];ReturnValue;taint",
         "java.lang;String;false;valueOf;(char);;Argument[0];ReturnValue;taint",
         "java.lang;String;false;valueOf;(char[],int,int);;Argument[0];ReturnValue;taint",
         "java.lang;String;false;valueOf;(char[]);;Argument[0];ReturnValue;taint",
+        // TODO: Should `append` and `write` be modelled for Appendable and Writer instead?
+        // Could then remove some of the modelled `append` method here and for StringBuilder
```
Comment (Marcono1234) [inline java/ql/lib/semmle/code/java/frameworks/Strings.qll:55]: Should `Appendable` and `Writer` be modelled?

File: java/ql/lib/semmle/code/java/frameworks/Strings.qll [hunk 4]
```diff
@@ -35,26 +44,43 @@ private class StringSummaryCsv extends SummaryModelCsv {
...
-        "java.lang;AbstractStringBuilder;true;toString;;;Argument[-1];ReturnValue;taint",
+        "java.lang;AbstractStringBuilder;true;replace;;;Argument[-1];ReturnValue;value",
+        "java.lang;AbstractStringBuilder;true;reverse;;;Argument[-1];ReturnValue;value",
+        "java.lang;AbstractStringBuilder;true;substring;;;Argument[-1];ReturnValue;value",
         "java.lang;StringBuffer;true;StringBuffer;(CharSequence);;Argument[0];Argument[-1];taint",
         "java.lang;StringBuffer;true;StringBuffer;(String);;Argument[0];Argument[-1];taint",
-        "java.lang;StringBuilder;true;StringBuilder;;;Argument[0];Argument[-1];taint",
-        "java.lang;CharSequence;true;subSequence;;;Argument[-1];ReturnValue;taint"
+        "java.lang;StringBuilder;true;StringBuilder;(CharSequence);;Argument[0];Argument[-1];taint",
+        "java.lang;StringBuilder;true;StringBuilder;(String);;Argument[0];Argument[-1];taint",
```
Comment (Marcono1234) [inline java/ql/lib/semmle/code/java/frameworks/Strings.qll:79]: Have specified the parameter types here to avoid tracking `StringBuilder(int)` where the argument only represents the capacitiy.

File: java/ql/lib/semmle/code/java/frameworks/Strings.qll (hunk 2 above)
Comment (smowton) [reply java/ql/lib/semmle/code/java/frameworks/Strings.qll:26]: Because `TaintTracking::Configuration` specifies an `allowImplicitRead` method that allows implicit reads from an array, collection or map-value (i.e., we can treat `y = propagateTaint(x)` or `sinkTaint(x)` as if they were `y = propagateTaint(x[n])` or `sinkTaint(x.get(n))`), in practice it doesn't make much difference for typical taint-tracking applications. In addition it specifies an `isAdditionalFlowStep` implementation that provides the opposite blurring of taint, saying that reading from a tainted map or tainted array yields tainted content.

In summary, for taint-tracking purposes `Field` and `SyntheticField` content continue to be faithfully distinguished, but array, collection and map content are blurred in both directions, with a tainted whole implying a tainted element and vice versa. You can specify them for clarity in the CSV description, but it won't make much (any?) difference to your results.

File: java/ql/lib/semmle/code/java/frameworks/Strings.qll (hunk 3 above)
Comment (smowton) [reply java/ql/lib/semmle/code/java/frameworks/Strings.qll:55]: This sounds like a positive step to me

File: java/ql/lib/semmle/code/java/frameworks/Strings.qll [hunk 5]
```diff
@@ -11,19 +11,28 @@ private class StringSummaryCsv extends SummaryModelCsv {
...
-        "java.lang;String;false;join;;;Argument[0..1];ReturnValue;taint",
+        "java.lang;String;false;join;;;Argument[0..1];ReturnValue;taint", // TODO: ArrayElement of Argument?
+        "java.lang;String;false;lines;;;Argument[-1];Element of ReturnValue;taint",
         "java.lang;String;false;repeat;(int);;Argument[-1];ReturnValue;taint",
+        "java.lang;String;false;replace;;;Argument[1];ReturnValue;taint",
+        "java.lang;String;false;replace;;;Argument[-1];ReturnValue;taint",
+        "java.lang;String;false;replaceAll;;;Argument[1];ReturnValue;taint",
+        "java.lang;String;false;replaceAll;;;Argument[-1];ReturnValue;taint",
+        "java.lang;String;false;replaceFirst;;;Argument[1];ReturnValue;taint",
+        "java.lang;String;false;replaceFirst;;;Argument[-1];ReturnValue;taint",
```
Comment (smowton) [inline java/ql/lib/semmle/code/java/frameworks/Strings.qll:34]: These ones have non-CSV models that should be cleaned up concurrently (search for `"replaceAll"` for example)

**Review by smowton — COMMENTED** [review]
There's few enough new methods here that it's probably best to just write some tests by hand for these, particularly since there are lambda tests involved and AFAIK the auto test generation stuff can't do that yet.

Suggest writing the tests then running them without this in place, as a quick way to find string methods that have non-CSV models that can be cleaned up.
